contracts/src/utils/structs/checkpoints.rs

```diff
@@ -0,0 +1,463 @@
+        let pos = self.length();
+        if pos > U256::ZERO {
+            let last = self._access(pos - uint!(1_U256));
+            let last_key = last._key.get();
+            let last_value = last._value.get();
+
+            // Checkpoint keys must be non-decreasing.
+            if last_key > key {
+                return Err(CheckpointUnorderedInsertion {}.into());
+            }
+
+            // Update or push new checkpoint
+            if last_key == key {
+                self._access_mut(pos - uint!(1_U256))._value.set(value);
+            } else {
+                self._unchecked_push(key, value);
+            }
+            Ok((last_value, value))
+        } else {
+            self._unchecked_push(key, value);
+            Ok((U160::ZERO, value))
+        }
+    }
+
+    /// Return the index of the last (most recent) checkpoint with key
+    /// lower or equal than the search key, or `high` if there is none.
+    /// Indexes `low` and `high` define a section where to do the search, with
+    /// inclusive `low` and exclusive `high`.
+    ///
+    /// WARNING: `high` should not be greater than the array's length.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the checkpoint's state.
+    /// * `key` - Checkpoint key to lookup.
+    /// * `low` - inclusive index where search begins.
+    /// * `high` - exclusive index where search ends.
+    fn _upper_binary_lookup(
+        &self,
+        key: U96,
+        mut low: U256,
+        mut high: U256,
+    ) -> U256 {
+        while low < high {
+            let mid = low.average(high);
+            if self._access(mid)._key.get() > key {
+                high = mid;
+            } else {
+                low = mid + uint!(1_U256);
+            }
+        }
+        high
+    }
+
+    /// Return the index of the first (oldest) checkpoint with key is
+    /// greater or equal than the search key, or `high` if there is none.
+    /// Indexes `low` and `high` define a section where to do the search, with
+    /// inclusive `low` and exclusive `high`.
+    ///
+    /// WARNING: `high` should not be greater than the array's length.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the checkpoint's state.
+    /// * `key` - Checkpoint key to lookup.
+    /// * `low` - inclusive index where search begins.
+    /// * `high` - exclusive index where search ends.
+    fn _lower_binary_lookup(
+        &self,
+        key: U96,
+        mut low: U256,
+        mut high: U256,
+    ) -> U256 {
+        while low < high {
+            let mid = low.average(high);
+            if self._access(mid)._key.get() < key {
+                low = mid + uint!(1_U256);
+            } else {
+                high = mid;
+            }
+        }
+        high
+    }
+
+    /// Immutable access on an element of the checkpoint's array.
+    /// The position is assumed to be within bounds.
+    /// Panic when out of bounds.
```

Comment:
I think we should have `Panics` section like here:
https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/contracts/src/token/erc20/mod.rs#L324-L326